codec-redis/src/main/java/io/netty/handler/codec/redis/RedisDecoder.java

```diff
@@ -0,0 +1,253 @@
+ * {@link RedisArrayAggregator} or processed directly.
+ */
+public class RedisDecoder extends ByteToMessageDecoder {
+
+    private static final int CRLF_LENGTH = 2;
+    private static final int NULL_RESPONSE_LENGTH = -1;
+    private static final int REDIS_MESSAGE_MAX_LENGTH = 512 * 1024 * 1024; // 512MB
+    private static final int POSITIVE_LONG_MAX_LENGTH = 19; // length of Long.MAX_VALUE
+    private static final short EOL = PlatformDependent.BIG_ENDIAN_NATIVE_ORDER
+            ? (short) (((int) '\n' << 8) | '\r')
+            : (short) (((int) '\r' << 8) | '\n');
+
+    private final ToPositiveLongProcessor toPositiveLongProcessor = new ToPositiveLongProcessor();
+
+    // current decoding states
+    private State state = State.DECODE_TYPE;
+    private RedisMessageType type;
+
+    // for decoding bulk string
+    private int bulkStringLength;
+
+    private enum State {
+        DECODE_TYPE,
+        DECODE_INLINE, // SIMPLE_STRING, ERROR, INTEGER
+        DECODE_LENGTH, // BULK_STRING, ARRAY_HEADER
+        DECODE_BULK_STRING,
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        try {
+            for (;;) {
+                switch (state) {
+                case DECODE_TYPE:
+                    if (!decodeType(in)) {
+                        return;
+                    }
+                    break;
+                case DECODE_INLINE:
+                    if (!decodeInline(in, out)) {
+                        return;
+                    }
+                    break;
+                case DECODE_LENGTH:
+                    if (!decodeLength(in, out)) {
+                        return;
+                    }
+                    break;
+                case DECODE_BULK_STRING:
+                    if (!decodeBulkString(in, out)) {
+                        return;
+                    }
+                    break;
+                default:
+                    throw new RedisCodecException("Unknown state: " + state);
+                }
+            }
+        } catch (RedisCodecException e) {
+            resetDecoder();
+            throw e;
+        } catch (Exception e) {
+            resetDecoder();
+            throw new RedisCodecException(e);
+        }
+    }
+
+    private void resetDecoder() {
+        state = State.DECODE_TYPE;
+        bulkStringLength = 0;
+    }
+
+    private boolean decodeType(ByteBuf in) throws Exception {
+        if (!in.isReadable()) {
+            return false;
+        }
+        type = RedisMessageType.valueOf(in.readByte());
+        state = type.isInline() ? State.DECODE_INLINE : State.DECODE_LENGTH;
+        return true;
+    }
+
+    private boolean decodeInline(ByteBuf in, List<Object> out) throws Exception {
+        ByteBuf lineBytes = readLine(in);
+        if (lineBytes == null) {
+            if (in.readableBytes() > REDIS_MESSAGE_MAX_LENGTH) {
```

Comment:
@jongyeol - Sounds like a good default. However I'm not sure what other versions of redis use, and it is not part of the protocol specification, perhpas it is best to allow it to be configurable via the constructor too?